src/main/java/org/zalando/nakadi/repository/MultiTimelineEventConsumer.java

```diff
@@ -0,0 +1,264 @@
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.nakadi.domain.ConsumedEvent;
+import org.zalando.nakadi.domain.EventTypePartition;
+import org.zalando.nakadi.domain.NakadiCursor;
+import org.zalando.nakadi.domain.PartitionStatistics;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.domain.TopicPartition;
+import org.zalando.nakadi.exceptions.InvalidCursorException;
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.service.timeline.TimelineService;
+import org.zalando.nakadi.service.timeline.TimelineSync;
+import org.zalando.nakadi.util.NakadiCollectionUtils;
+
+public class MultiTimelineEventConsumer implements EventConsumer {
+    private final String clientId;
+    private final Map<EventTypePartition, NakadiCursor> latestOffsets = new HashMap<>();
+    private final Map<String, List<Timeline>> consumedEventTypes = new HashMap<>();
+    private final Map<String, TimelineSync.ListenerRegistration> timelineRefreshListeners = new HashMap<>();
+    private final Map<TopicRepository, EventConsumer> eventConsumers = new HashMap<>();
+    private final LinkedList<ConsumedEvent> eventsQueue = new LinkedList<>();
+    private final Map<EventTypePartition, String> borderOffsets = new HashMap<>();
+    private final TimelineService timelineService;
+    private final TimelineSync timelineSync;
+    private final BlockingQueue<Runnable> queuedTasks = new LinkedBlockingQueue<>();
+    private static final Logger LOG = LoggerFactory.getLogger(MultiTimelineEventConsumer.class);
+
+    public MultiTimelineEventConsumer(
+            final String clientId,
+            final TimelineService timelineService,
+            final TimelineSync timelineSync) {
+        this.clientId = clientId;
+        this.timelineService = timelineService;
+        this.timelineSync = timelineSync;
+    }
+
+    @Override
+    public Set<TopicPartition> getAssignment() {
+        return latestOffsets.values().stream().map(NakadiCursor::getTopicPartition).collect(Collectors.toSet());
+    }
+
+
+    @Override
+    public Optional<ConsumedEvent> readEvent() {
+        Runnable task;
+        while ((task = queuedTasks.poll()) != null) {
+            task.run();
+        }
+        final Optional<ConsumedEvent> result = poll();
+        if (result.isPresent()) {
+            final NakadiCursor position = result.get().getPosition();
+            final EventTypePartition etp = position.getEventTypePartition();
+            latestOffsets.put(etp, position);
+            final String borderOffset = borderOffsets.get(etp);
+            if (null != borderOffset && borderOffset.compareTo(position.getOffset()) <= 0) {
```

Comment:
Does it mean that when reading, the consumer reached the limit of available events? I find it useful to extract the result of the expression to a variable with a good explanatory name or add a comment like: "switch to a new timeline when all events from the previous timeline have been consumed".